# Verifying Sashimi plot filtering

Test verified ...

 **Prerequisites**:

 - Dataset1
 - BAM track (***brain_th.bam***) that contains splice junctions tracks in Dataset1.
 - Global Settings >> **Alignments** tab >> **Splice Junctions** section >> **Enable filtering** checkbox is unshecked.

| Steps | Actions | Expected results |
| :---: | --- | --- |
| 1 | Login to NGB | |
| 2 | Go to Dateset panel. Expand Dataset1 using an arrow icon next to a group name. | |
| 3 | Tick track1.bam in the Dataset1 list to toggle its visibility. | |
| 4 | Select needed chromosome from the chromosome selector dropdown (zoom in track if needed). | |
| 5 | Open **General** menu in BAM track's header. | |
| 6 | Click **Sashimi plot** option. | **Sashimi plot**  popup opens and contains: <li>GENE reference. <li> BAM track itself with the visualization of coverage and splice junctions that visualized only by curved lines.<li>Splice junction depth value is displayed near each junction.|
| 7 | Look at the depth value for junctions. | |
| 8 | Open Global Settings. | |
| 9 | Go to **Alignments** tab. | |
| 10 | Navigate to **Splice Junctions** section. | |
| 11 | Check **Enable filtering** checkbox. | |
| 12 | Enter average value from step 7 into the minimum coverage threshold value field. Click **Save** button. | |
| 13 | Repeat steps 5-6. |**Sashimi plot**  popup opens: <li> Only splice junctions with depth less than threshold aren't shown.<li> Configured threshold is displayed at BAM track header as label  `Splice junctions minimum coverage: <value>`. |
| 14 | Repeat steps 8-10. | |
| 15 | Uncheck **Enable filtering** checkbox. Click **Save** button. | |
| 16 | Repeat steps 5-6. |<li>All splice junctions are shown.<li>Label `Splice junctions minimum coverage: <value>` isn't shown. |