# Motifs search from read (BAM)

Test verifies

- Motifs search from BAM track, displaying results in new MOTIFS panel and on tracks (positive and negative strands)

**Prerequisites:**
- dataset = **sample1**, .bam = **sv_sample1.bam**
- BAM sequence: `GGTTCATCTCACAGGGGCTTGTCAGACAGTAGGTGCAGCCCATGGAGCAGGGCGGGGCATCGCCTCACCCAGGA`

| Steps | Actions                                                                                | Expected results                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|:-----:|----------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|   1   | Login to NGB                                                                           |                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|   2   | Go to **Datasets** panel                                                               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|   3   | Select dataset, bam file from **Prerequisites**                                        | BAM file **sv_sample1.bam** is selected in dataset (**sample1**)                                                                                                                                                                                                                                                                                                                                                                                                          |
|   4   | Go to **Browser** panel                                                                |                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|   5   | Go to Coordinates and search input control at right of tab's header                    |                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|   6   | Set coordinates  **1: 61943246 - 61943319** and click **Enter**                        | **BAM** track displays in the **Browser** panel                                                                                                                                                                                                                                                                                                                                                                                                                           |
|   7   | At the bam-track, select **Collapsed** view (if it was not set previously)             |                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|   8   | Click on BAM read                                                                      | <li> Context menu is displayed <li>**Motifs search** displays between **Open mate region in split view** and **BLAT search**                                                                                                                                                                                                                                                                                                                                              |
|   9   | Click on **Motifs search** in the context menu                                         | **Search motifs** pop-up is appeared with following fields and buttons: <li>**Pattern** (*mandatory*) field - filled by the selected read sequence by default (sequence displays in **Prerequisites**) <li>**Title** (*optional*) field - empty by default <li>**Search whole reference**. ***Warning: this can take a long a time*** (*optional*) checkbox - unset by default <br><li>**SEARCH** button (is enabled) <li>**CANCEL** button <li>**x** (close) button      |
|  10   | Fill the **Title** field (e.g. BAM)                                                    |                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|  11   | Click **SEARCH** button                                                                | <li>**Motifs search** pop-up is closed <li>**MOTIFS** panel with details is displayed first at the right side in the additional panels <li>The search process is displayed at the top of the panel as a loader with the heading **Loading results**                                                                                                                                                                                                                       |
|  12   | Wait until the search process is completed                                             | The loader is disappeared <br>The following values display in the panel: <li> Motif's name specified as **Title** at step 10 is displayed at the top of the panel <li> **<** - button displays to the left of the Motif's name (returns to list of all search motifs) <li>A table with the specific search result details (all found motif matches) with columns: **Chromosome, Start, End, Strand, Gene** <li>Search results are sorted ascending by *Start* column <li> |
|  13   | Look at values in the table                                                            | Motif search results are displayed in the table: <br><table><thead><tr><th>Chr.</th><th>Start </th><th>End</th><th>Strand </th><th>Gene</th></tr><tr><td>1 </td><td>61943237 </td><td>61943310 </td><td>+</td><td>INADL </td></tr></table> <ul><li> Filter for **Chromosome** column is disabled with **1** value. Filters for **Strand** and **Gene** columns are empty                                                                                                  |
|  14   | Go to **BROWSER** panel                                                                | Two tracks are displayed in the top of panel (under the **REFERENCE** track): <li>`<TITLE>_POSITIVE` <li> `<TITLE>_NEGATIVE`, <br> where `<TITLE>` is the search task *Title* specified at step 10 (*BAM_POSITIVE* and *BAM_NEGATIVE*)  <ul>***General** context menu* and ***Navigation** with  **Prev** and **Next** buttons* are displayed in the both track's header                                                                                                  |
|  15   | Look at `<TITLE>_POSITIVE` track                                                       | <li> The Sequence matching the specified search pattern is displayed on the track <li> The sequence is shown as colorized rectangle (blue color by default) <li> The rectangle has white arrows inside in the direction of the positive strand (left to right) 	                                                                                                                                                                                                          |
|  16   | Look at `<TITLE>_NEGATIVE` track                                                       | Track is empty (doesn't contain the sequence matching) <ul> *In the case when there is a match in sequence: <br/>  - a rectangle is shown in red color by default <br/> - a rectangle has white arrows inside in the reverse direction of the positive strand (right to left)*                                                                                                                                                                                            |
|  17   | Repeat steps 7-15 for **Expanded** view                                                | All opened motifs results tracks in **Browser** remain opened when starting a new search                                                                                                                                                                                                                                                                                                                                                                                  |
|  18   | Click **<** button                                                                     | <li> A table with all runned searches for the current dataset is displayed with following columns: <ul><li>**Name** - displays search task titles specified at step 10 for each case <li>**Motif** - displays search patterns for each case <li>**Search type** - displays search type *Chromosome* </ul><li> 2 search rows are displayed in the table <li> search results are sorted in order of appearance (the last search is shown on top)                            |
|  19   | Step by step click results from **MOTIFS** panel and click **<** button to return back | <li> The corresponding chromosome and coordinates are opened <li> The corresponding motifs tracks are re-opened with results on them <li> For each new successful motifs search, additional two corresponding tracks are opened                                                                                                                                                                                                                                           |


